handler/rfc8628/auth_handler.go

```diff
@@ -55,3 +52,49 @@ func (d *DeviceAuthHandler) HandleDeviceEndpointRequest(ctx context.Context, dar
 	resp.SetInterval(int(d.Config.GetDeviceAuthTokenPollingInterval(ctx).Seconds()))
 	return nil
 }
+
+func (d *DeviceAuthHandler) handleDeviceCode(ctx context.Context, dar fosite.DeviceRequester) (string, error) {
+	code, signature, err := d.Strategy.GenerateDeviceCode(ctx)
+	if err != nil {
+		return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	dar.GetSession().SetExpiresAt(fosite.DeviceCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)))
+	if err = d.Storage.CreateDeviceCodeSession(ctx, signature, dar.Sanitize(nil)); err != nil {
+		return "", errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	return code, nil
+}
+
+func (d *DeviceAuthHandler) handleUserCode(ctx context.Context, dar fosite.DeviceRequester) (string, error) {
+	userCode, err := retry.DoWithData(
+		func() (string, error) {
+			code, signature, err := d.Strategy.GenerateUserCode(ctx)
+			if err != nil {
+				return "", err
+			}
+
+			dar.GetSession().SetExpiresAt(fosite.UserCode, time.Now().UTC().Add(d.Config.GetDeviceAndUserCodeLifespan(ctx)).Round(time.Second))
+			err = d.Storage.CreateUserCodeSession(ctx, signature, dar.Sanitize(nil))
+			if err != nil {
+				return "", err
+			}
+
+			return code, nil
+		},
+		retry.Attempts(3),
+		retry.Context(ctx),
+		retry.Delay(0),
+		retry.DelayType(retry.CombineDelay(retry.FixedDelay)),
+		retry.LastErrorOnly(true),
+		retry.RetryIf(func(err error) bool {
+			return errors.Is(err, sqlcon.ErrUniqueViolation)
```

Comment:
I don't like this introduction of the error from `sqlcon`.

I think it's better to introduce a new error in `errors.go` and use that here. And the `CreateUserCodeSession` method in Hydra will explicitly throw such an error when a `sqlcon.ErrUniqueViolation` error is thrown by `createSession` method.

Let me know what you think.

Reply:
I think this sounds better. Even though sqlcon comes from an ory project as well, I would prefer the retry logic to be defined entirely on fosite.

Reply:
I d say we don't care, if u look at `handleDevice` we pass up the error, in this case what I would do is `continue` in the case of a SQLcon error up to x times and then bubble it up, if we error at any other point we bubble it up directly 

we won't need to know it s a SQLcon error but just know that it comes from that storage func call